ranger/core/main.py

```diff
@@ -161,6 +161,7 @@ def main(
                     with open(tabs_datapath, 'r', encoding="utf-8") as fobj:
                         tabs_saved = fobj.read().partition('\0\0')
                         fm.start_paths += tabs_saved[0].split('\0')
+                        fm.start_paths = list(filter(os.path.isdir, dict.fromkeys(fm.start_paths)))
```

Comment:
I don't think dictionaries actually guarantee any ordering in the keys? As eMPee584 points out maintaining the ordering is desirable.

Also, can we maybe notify the user of any paths that are filtered out for no longer existing? It's possible a user forgot about mounting a partition and doesn't necessarily want to lose the saved tab. So being informed of tabs being removed and for which paths would be nice.

Reply:
Actually, maybe we can even keep the tabs for non-existent directories. If we can keep Ranger from crashing when switching to tabs for which the path doesn't exist. That way you could keep tabs open dedicated to network volumes or removable drives.

Reply:
> maintaining the ordering is desirable

Maintain tab order while filtering out duplicates and dead entries

Keep only the first occurrence of a tab with a duplicate path, if a tab with the current launch path already exists, switch to that tab after launch instead of creating a new one at the top of the list.
34e813e8243bb8f7c42bf32799b7fb13ecd9a10e

Reply:
Regarding keeping tabs with non-existent paths (unmounted network volumes etc.), what should we do with them when switching tabs? Just skip it and go directly to the next one? Ok, but what about deleted dirs, which never get back again? How should we manage them? Of course, we can switch to tab and if its path doesn't exist go to some default one (user's $HOME for example), but what is the point if we destroy all these tabs with unmounted volumes this way?

There are many possibilities to revert to paths that didn't exist at startup if they become available again (built-in bookmarks, [fzf-marks](https://github.com/laggardkernel/ranger-fzf-marks), [fasd](https://github.com/gotbletu/shownotes/blob/master/ranger_fasd_fzf.md), [cd .. keymaps](https://github.com/ranger/ranger/blob/master/ranger/config/rc.conf?plain=1#L450-L466) for favorite directories, maybe something else...).
I mean, if we want to keep things simple, just clearing tabs with "dead" paths on startup is a good enough solution, IMO.

Reply:
Ah, interesting. I missed the change to dict, used to be ordered by the hashes of the objects if I'm not mistaken. We do support older versions than Python 3.6 but let's not have that hold us back in this instance and just add a note.

My desired behavior re tabs with unavailable paths is they simply display the usual "empty" or whatever placeholder Ranger always uses and not interfere with moving across tabs or moving tabs.
Testing on master puts dead tabs in the first still existing parent AFAICS, so we could change this to reflect the original path and show "empty" in the parents if they don't exist. But since it doesn't cause crashes it's OK if you'd rather leave this unaltered.

Removing duplicate paths isn't actually a good idea IMO, now that I've thought about it more. It can be totally valid for someone to always want multiple tabs that are in the same directory. I think the main issue is we always create a new tab when starting ranger? We could suppress doing so if the path is already open in a tab and move to that tab instead. We do need to consider multiple paths passed on the command line. If I pass multiple paths I do expect them to open in the order I pass them. If you already have them open in saved tabs they might be in a different order, we could look for the greatest overlap and then add missing tabs on either end but that seems fairly complicated.

Actually I noticed that Ranger seems to only restore tabs when no paths are passed on the command line so that may not be a problem. I'm not sure that's desirable behavior though, should we restore tabs even when paths are passed on the command line?

To avoid the duplicate directories, maybe we simply want to move to the PWD if it is already an open tab and no path is passed?

Don't feel a need to implement code whenever I say something BTW, let's discuss what we really want first to avoid extra work on your part.

Reply:
> let's discuss what we really want first

Hello! Yes, sure! I completely agree, and let's discuss. I just got some free time this weekend, and I'm enjoying improving my favorite console FM a bit. 😄

> Removing duplicate paths isn't actually a good idea IMO

It doesn't make any sense to me personally, but ok, I agree that duplicating tabs at startup can be useful to someone (although the content of these tabs will be exactly the same, I mean sorting, cursor position ... everything). If you really need a copy of the tab, you can just press Ctrl+n as soon as you switch to that tab and that's it. But again, I'm totally OK with keeping duplicates on startup, but maybe we can do that optional? Personally, I'm much more annoyed by switching and deleting duplicate tabs from a previous session than making a copy of a tab when I need it. For some, it might be the other way around. So maybe the user has to decide on this part of the startup approach?

> I think the main issue is we always create a new tab when starting ranger? We could suppress doing so if the path is already open in a tab and move to that tab instead.

In fact, this part is already implemented, so regardless of the optional deduplication at startup, we can already switch to the PWD tab if it already exists, instead of creating a new one (or switch to the first instance of existing tabs if we keep duplicates and have this tab more than once at startup).

> should we restore tabs even when paths are passed on the command line?

I don't think so, currently, we ignore everything else if we specify certain paths that we want to have in tabs on startup and I think that makes sense. However, it's up to you to decide, because it's absolutely not my use case. I never start Ranger with any arguments, but always just start it from PWD, get my saved tabs, and jump to where I need to.

> Testing on master puts dead tabs in the first still existing parent AFAICS, so we could change this to reflect the original path and show "empty" in the parents if they don't exist.

Maybe I didn't understand you completely, but it makes no sense for me to let the user switch to a tab with a "temporarily" unavailable path and show him "empty" or any other placeholder instead of the content of the desired dir. Switching to it and immediately jumping from there to the closest existing parent or any other path makes no sense as well, IMO. Because if we decided to keep tabs with unavailable paths to use them when these paths become available again, we don't really want to lose these tabs this way. So I still think that the best approach is to silently skip these tabs until they are "resurrected", or (optionally) ask the user whether to delete such a tab (if its path is never expected to be available again, for example).

Reply:
P.S.:
It just occurred to me: imagine that the user in charge of the annual financial report for a very important audit, on the eve of a major transaction for the company that depends on this report, switches to the network drive tab where this report is stored, which suddenly turned out to be unmounted from the file system and sees there `empty `. Who will take responsibility for his heart attack? :laughing:

Reply:
Current behavior doesn't need that much changing I think. You can see what I'd expect on a dead tab by navigating somewhere in Ranger and then deleting a parent directory in another shell. Then run `:reset` in Ranger. It should show `empty` for the parent and `not accessible` for the current directory. The only part that could bear improvement IMO is that `:cd ..` or `h` will change to the first accessible parent instead of a single step of the path. But that change can be left for another PR.

The setting isn't good enough because the "safe" option, i.e., not simply deleting the tabs, *requires* attentive interaction every time you move over the tab, and you need to answer correctly every single time. It's not hard to accidentally hit `y` when you really meant `n`. Brainfarts happen.

That scenario is actually a vote in favor of what I'm suggesting. I'd rather realize the official copy isn't accessible right now, rather than accidentally skip to the next tab that has something potentially similar open and risk not realizing I'm not actually working on the important document.

If you're not interested in implementing this, that's absolutely fine and you can cut the PR down to just avoiding duplicate creation.

Reply:
I have made several attempts to implement the described approach, and the best I have come up with is the following behavior: 

Suppose we have some directory open in a tab:
![image](https://user-images.githubusercontent.com/63558838/229713477-9fb91fd8-bf9a-4d2c-9507-b81e0a584b51.png)
![image](https://user-images.githubusercontent.com/63558838/229713594-c2e0b4cd-7e1b-431f-b0c7-ec804a2d984a.png)

Now we remove this directory from another tab and return to this tab:
![image](https://user-images.githubusercontent.com/63558838/229713807-7cf70561-76ba-4ffa-9891-2b4fcab94bdd.png)
![image](https://user-images.githubusercontent.com/63558838/229713897-c602e421-3e08-4c9e-befa-b8b882125133.png)
We have received the desired `not accessible` message.

Here's what it looks like with a longer nested directory path:
![image](https://user-images.githubusercontent.com/63558838/229714100-d9d4d63d-28de-43a8-920a-db40c30fdd3c.png)
![image](https://user-images.githubusercontent.com/63558838/229714551-a3cbeae8-0fbe-4c73-92ff-c287f4804fe2.png)
![image](https://user-images.githubusercontent.com/63558838/229714698-fe61514d-b7d8-425c-b424-2e66485396bd.png)
![image](https://user-images.githubusercontent.com/63558838/229714792-1f6f14ea-6f6f-472c-ace2-96b3ef5ed9ec.png)

If we create this directory again, the tab will come to life again:
![image](https://user-images.githubusercontent.com/63558838/229715404-5eeb8275-0afd-406b-9028-80a8868816dd.png)
![image](https://user-images.githubusercontent.com/63558838/229715480-0bb05159-c52d-485f-b310-8588dcfbb8ff.png)

So far everything looks good and I was almost happy with this "fix", but then I noticed some shortcomings.

1. If we launch Ranger in a state where the open tab path is "dead" and switch to that tab, we see only empty space with no columns with the message "not available".
![image](https://user-images.githubusercontent.com/63558838/229716775-60498e29-0b79-42c8-863e-8e582ddc4c3f.png)
2. If we try to create any file or directory in this "dead" tab, it will be created in the path of the previous "live" tab.
![image](https://user-images.githubusercontent.com/63558838/229717379-22f59dfb-6692-48c8-83ac-354cc8475672.png)
![image](https://user-images.githubusercontent.com/63558838/229717554-07f4679d-556c-4815-982b-c60da6d92443.png)

I tried to solve these problems many times over several days without success.
It seems my approach is not the correct way how we should achieve this.
I've pushed this WIP commit, in case you're interested in taking a look and suggesting how to fix it.
Otherwise, I will cut down this commit and leave only getting rid of the duplicate existing tab at startup.

Reply:
> Would you be interested in looking into this further?

Well, yes and no. I have already made several attempts to somehow improve what you saw here. But no luck, unfortunately. And I think I don't have enough free time to continue like this (reverse coding blindly). As I said before, I'm not a very experienced Python developer and not entirely familiar with the entire Ranger codebase, and I'm stuck. Therefore, I'm afraid that if no one supports me in this matter, I will not continue this fix, sorry.
If anyone else is interested in this later and decides to continue from now on, I can fork and keep these changes in another branch of my fork of this repo or even create another issue with that branch. But cut off everything but duplicate tabs on startup fix in current branch, so you can easily merge it.

Reply:
Totally understandable. No need to keep this in a fork though. I intend to merge it regardless.